clickhouse-core/src/main/scala/xenon/clickhouse/hash/HashFunc.scala

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xenon.clickhouse.hash
+
+import scala.reflect.ClassTag
+
+abstract class HashFunc[T: ClassTag] {
```

Comment:
I suppose the result of the hash function should always be Long, any other cases?

Reply:
murmurHash2/3 in 32bit returns UInt32 in Clickhouse, and we are casting it into Int64 with UInt32 range. For strictly type matching, we should make them return Java's `int`, but `int` requires UInt32 range handling, especially when we cast it into Int64 afterwards. Also, Clickhouse has `murmurHash3_128` and `sipHash128` that returns UInt128 (as fixed length string). I am leaving this template for future support of this case.

Reply:
Educated, the ClickHouse's hash functions are more complex than the Java world.

Reply:
Updated 32bit hash function returns Int, after fixing the test issue and finding some problems with the former code. I think this will make the UInt32 range handling cleaner.